---
title: "BCB420 Assignment"
author: "Laeticia Liawas"
date: "2026-01-27"
output:
  html_document:
    toc: true                  
    toc_depth: 3                
    toc_float: true             
    code_folding: hide          
bibliography: references.bib    
link-citations: true            
nocite: '@*' # cite all, even if not present as an in-text citation
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE, # suppressing warning messages
                      message = FALSE,
                      knitr.kable.max_rows = 6)

# required packages
```

# Introduction {#introduction}

-> what you are investigating and why
-> background to understand what the dataset is measuring?
-> assume everyone knows RNAseq, and the concept of a regular RNAseq pipeline.

-> problem statement 
dengue fever, why it is dangerous/need to be studied. 
hILC, what they are. 
why connect to dengue fever.
-> how this analysis can connect to the problem.

# Selecting an expression dataset {#selecting-data-set}

In this section, I will detail my own rationale behind my choice of dataset. In this section, I will also go through steps to download the data.

## How the dataset was chosen {#ds-choice-rationale}

Being from a tropical country myself, the dangers of dengue fever is one not completely unknown to me. For a disease which has led to countless warning infomercials being broadcasted on TV since I was a kid, I was surprised to know that we still don't know everything about this disease. 

## Downloading the dataset {#ds-download}

# Cleaning the data and mapping to HUGO symbols {#cleaning-n-mapping}

This section pertains to showing overview statistics for the dataset, discussing how potential

## Overview statistics {#overview-stats}

## Outliers {#outliers}

The original paper did not mention any details regarding any instrument errors that may lead to outliers. However, in their analysis the authors did mention that they excluded genes for which all samples had 0 expression. Thus, I decided to do the same in my analysis.

## Identity mapping {#mapping}

## Normalisation {#normalise}

# Differential gene expression {#deg}

# Questions {#questions}
